File: ios/GrMob/App/AppWindow.swift
import SwiftUI

/// The iOS half of grmob's window event: tells Go how big the app window is,
/// through the "window" host event (core/window.go).
///
///     GeometryReader size          ──▶ { width, height }   (points)
///     GeometryReader safeAreaInsets ──▶ { insets: {…} }
///
/// No fold is ever reported. No iPhone or iPad has a hinge, so the payload
/// omits the "fold" key, which is exactly what core reads as "no fold". What
/// iOS *does* have is a window that changes size under a running app —
/// Split View, Slide Over and Stage Manager on iPad, rotation everywhere —
/// and that is the same size-class question a foldable asks, answered by the
/// same record, so a layout written for an unfolding phone adapts to an iPad
/// split without a line of platform code.
///
/// # Why a GeometryReader and not UIScreen
///
/// UIScreen.main.bounds is the display, and an iPad app in Split View has a
/// fraction of it. The window is what the tree is laid out in, and the size a
/// full-window view is offered is the window's. The reader sits in the root
/// view's background with the safe area ignored, so it measures the whole
/// window including the area under the status bar and home indicator — the
/// same window coordinates Android and the browser report in.
///
/// # The insets
///
/// Because the reader ignores the safe area, its proxy's `safeAreaInsets`
/// are the window's own — the status bar, the home indicator and, on a
/// notched phone in landscape, the sensor housing. They go to Go as
/// core.SafeInsets so a component can position itself against them; what
/// keeps ordinary content clear of the bars is still the SafeArea node, and
/// SwiftUI's own insetting behind it, not these numbers.
///
/// Go dedupes a size it already has, so SwiftUI re-offering the same size (it
/// does around scene connection) costs a bridge call and nothing more. The
/// insets ride in the same report and are deduped with it.
enum AppWindow {
    /// Main-actor because `GrMobRuntime.hostEvent` is, and because the
    /// callers — onAppear and onChange on the reader — already run there.
    @MainActor
    static func report(_ size: CGSize, insets: EdgeInsets, to runtime: GrMobRuntime) {
        // A zero size is the reader before layout, not a window; Go would
        // drop a zero-height report as meaningless anyway, but sending it
        // would briefly publish a zero-sized window to subscribers.
        guard size.width > 0, size.height > 0 else { return }
        // Serialized rather than interpolated, the same rule AppLifecycle
        // follows. The insets key below is what that bought: a nested object
        // added to the payload without touching how it is written, and
        // decoded by older core as absent rather than as a parse error.
        //
        // Physical edges, not leading/trailing: core.SafeInsets is defined
        // that way because the fold bounds beside it are, and a cutout is
        // where it is whatever the writing direction is. SwiftUI's EdgeInsets
        // is already physical here — the proxy reports the resolved values.
        let fields: [String: Any] = [
            "width": Double(size.width),
            "height": Double(size.height),
            "insets": [
                "top": Double(insets.top),
                "bottom": Double(insets.bottom),
                "left": Double(insets.leading),
                "right": Double(insets.trailing),
            ],
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: fields),
              let payload = String(data: data, encoding: .utf8)
        else { return }
        runtime.hostEvent("window", payload)
    }
}

/// The measuring view: a clear GeometryReader for GrMobRoot's background.
struct AppWindowReader: View {
    let runtime: GrMobRuntime

    var body: some View {
        GeometryReader { geo in
            Color.clear
                // onAppear for the first measurement, onChange for every
                // resize after it; onChange does not fire for the initial
                // value.
                .onAppear { AppWindow.report(geo.size, insets: geo.safeAreaInsets, to: runtime) }
                .onChange(of: geo.size) { _, size in
                    AppWindow.report(size, insets: geo.safeAreaInsets, to: runtime)
                }
                // The bars can move without the window resizing: a rotation
                // that puts the sensor housing on the other edge, a keyboard
                // that changes the bottom inset. onChange(of: size) alone
                // would miss those, so the insets are watched too.
                .onChange(of: geo.safeAreaInsets) { _, insets in
                    AppWindow.report(geo.size, insets: insets, to: runtime)
                }
        }
        .ignoresSafeArea()
    }
}

File: android/app/src/main/java/com/grmob/app/AppWindow.kt
package com.grmob.app

import androidx.activity.ComponentActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.window.layout.FoldingFeature
import androidx.window.layout.WindowInfoTracker
import androidx.window.layout.WindowLayoutInfo
import androidx.window.layout.WindowMetricsCalculator
import com.grmob.runtime.GrMobRuntime
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * The Android half of grmob's window event: tells Go how big the window is
 * and whether a fold crosses it, through the "window" host event
 * (core/window.go).
 *
 *   WindowMetricsCalculator ──▶ width, height          (px ÷ density → dp)
 *   FoldingFeature          ──▶ fold { state, orientation, separating,
 *                                      occluding, x, y, width, height }
 *   root WindowInsetsCompat ──▶ insets { top, bottom, left, right }
 *
 * # Why Jetpack WindowManager
 *
 * The platform has no fold API of its own below Android 15 — hinge
 * geometry lives in each OEM's window extensions — and WindowManager is the
 * library that normalizes them, including the emulator's foldable profiles.
 * On a device with no extensions (every non-foldable) windowLayoutInfo
 * simply emits no features, which is the "no fold" report.
 *
 * # When it reports
 *
 * windowLayoutInfo emits once when collection starts and again on every
 * change to the features: a hinge bending from flat to half-opened is an
 * emission with no configuration change. Folding or unfolding the device
 * *is* a configuration change (screenSize), so the Activity is recreated,
 * [attach] runs on the new instance and collection starts over — which is
 * also how a resize in multi-window mode arrives. Go dedupes a window it
 * already has (core.ReceiveWindow), so the recreation's repeat costs a
 * bridge call and nothing else.
 *
 * Collection is scoped to STARTED with repeatOnLifecycle: the tracker holds
 * a listener on the window extensions, and a stopped Activity has no window
 * a report could describe. Coming back to STARTED re-emits the current
 * layout, so a posture change made while backgrounded still lands.
 *
 * The insets have their own trigger, because they move without the fold or
 * the window doing anything — a rotation that carries the cutout to the
 * other edge, a gesture-nav bar that changes height. The natural hook,
 * setOnApplyWindowInsetsListener, is *not* used: it replaces a view's
 * inset handling rather than observing it, and on the decor view that is
 * the chain edge-to-edge and Compose's own WindowInsets depend on.
 * OnGlobalLayoutListener is additive, consumes nothing, and fires on the
 * layout pass an inset change causes anyway; the reported set is remembered
 * so the frequent calls turn into a comparison and nothing more.
 *
 * # Which insets
 *
 * systemBars + displayCutout, which is WindowInsets.safeDrawing minus the
 * IME — the same set Renderer.kt's SafeArea node applies, on purpose, so
 * the numbers Go reads describe the edge its own SafeArea keeps content
 * off. The keyboard is deliberately not in them: it is a transient overlay
 * with its own story (core/keyboard.go), not an edge of the window.
 *
 * # Units and coordinates
 *
 * Both the window bounds and FoldingFeature.bounds are in window pixels with
 * the origin at the window's top-left — under the status bar, since
 * MainActivity is edge to edge. Dividing by density puts them in dp, which
 * is the unit Go lays out in, and changes nothing else: core documents the
 * same window coordinate space for every host.
 */
object AppWindow {
    /**
     * Starts reporting for [activity]. Call from onCreate after the runtime
     * exists; the coroutine is tied to the Activity's lifecycle, so a
     * recreated Activity's collection replaces the destroyed one's rather
     * than stacking beside it.
     */
    fun attach(activity: ComponentActivity, runtime: GrMobRuntime) {
        val tracker = WindowInfoTracker.getOrCreate(activity)
        // The two triggers share what the other one knows. Both run on the
        // main thread — collect resumes on the lifecycleScope's Main
        // dispatcher and a layout pass is by definition on it — so plain
        // vars need no synchronization, and both die with this Activity
        // rather than living on the object across a recreation.
        var latest: WindowLayoutInfo? = null
        var reported: Insets? = null

        activity.lifecycleScope.launch {
            activity.repeatOnLifecycle(Lifecycle.State.STARTED) {
                tracker.windowLayoutInfo(activity).collect { info ->
                    latest = info
                    val now = currentInsets(activity)
                    reported = now
                    report(activity, runtime, info, now)
                }
            }
        }

        activity.window.decorView.viewTreeObserver.addOnGlobalLayoutListener {
            val now = currentInsets(activity)
            if (now == reported) return@addOnGlobalLayoutListener
            reported = now
            // latest is null only before the first emission, which is the
            // one case where the size may not be measured yet either; the
            // collector above will report both together in a moment.
            latest?.let { report(activity, runtime, it, now) }
        }
    }

    /** The window's safe-drawing edges in dp. See "Which insets" above. */
    private fun currentInsets(activity: ComponentActivity): Insets {
        val density = activity.resources.displayMetrics.density
        val root = ViewCompat.getRootWindowInsets(activity.window.decorView)
            ?: return Insets(0.0, 0.0, 0.0, 0.0)
        val i = root.getInsets(
            WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout(),
        )
        return Insets(
            top = i.top / density.toDouble(),
            bottom = i.bottom / density.toDouble(),
            left = i.left / density.toDouble(),
            right = i.right / density.toDouble(),
        )
    }

    /**
     * One report's insets, in dp. A data class so the "did they change?"
     * test above is a value comparison, which is the same thing core's
     * record does with the whole Window.
     */
    private data class Insets(
        val top: Double,
        val bottom: Double,
        val left: Double,
        val right: Double,
    )

    private fun report(
        activity: ComponentActivity,
        runtime: GrMobRuntime,
        info: WindowLayoutInfo,
        insets: Insets,
    ) {
        val density = activity.resources.displayMetrics.density
        // computeCurrentWindowMetrics, not the display's size: in split
        // screen or a freeform window the app has a fraction of the display,
        // and the window is what the tree is laid out in.
        val bounds = WindowMetricsCalculator.getOrCreate()
            .computeCurrentWindowMetrics(activity).bounds

        val payload = JSONObject()
            .put("width", bounds.width() / density)
            .put("height", bounds.height() / density)
            .put(
                "insets",
                JSONObject()
                    .put("top", insets.top)
                    .put("bottom", insets.bottom)
                    .put("left", insets.left)
                    .put("right", insets.right),
            )

        // At most one fold is reported. Every shipping foldable has one
        // hinge, and core's record has room for one; a device that someday
        // reports two gets its first, which still keeps First off a crease.
        val fold = info.displayFeatures.filterIsInstance<FoldingFeature>().firstOrNull()
        if (fold != null) {
            val b = fold.bounds
            payload.put(
                "fold",
                JSONObject()
                    .put("state", if (fold.state == FoldingFeature.State.HALF_OPENED) "half_opened" else "flat")
                    .put(
                        "orientation",
                        if (fold.orientation == FoldingFeature.Orientation.HORIZONTAL) "horizontal" else "vertical",
                    )
                    // isSeparating is WindowManager's own judgement — true
                    // for any half-opened hinge and for a flat seam between
                    // two panels — and is passed through rather than
                    // re-derived, so Go and Android agree on it.
                    .put("separating", fold.isSeparating)
                    .put("occluding", fold.occlusionType == FoldingFeature.OcclusionType.FULL)
                    .put("x", b.left / density)
                    .put("y", b.top / density)
                    .put("width", b.width() / density)
                    .put("height", b.height() / density),
            )
        }
        // hostEvent is safe from any thread and serializes with UI events on
        // the runtime's executor.
        runtime.hostEvent("window", payload.toString())
    }
}
